File: ts/project/.sdk/tm/swift/Sources/ProjectNameSDK/utility/Make.swift
// ProjectName SDK utility: the operation pipeline stages - point resolution,
// spec/url/fetchdef construction, transport request, and response/result
// shaping.

import Foundation

func makePointUtil(_ ctx: Context) throws -> VMap? {
  if let stored = ctx.out["point"], let sp = stored {
    // A PrePoint feature hook (e.g. rbac) may short-circuit the operation by
    // storing an error here; surface it before any endpoint resolution.
    if let err = sp as? Error { throw err }
    if let tm = sp as? VMap { ctx.point = tm; return tm }
  }

  let op = ctx.op!
  let options = ctx.options

  let allowOp = gpath(options, "allow", "op").asString ?? ""
  if !allowOp.contains(op.name) {
    throw ctx.makeError("point_op_allow",
      "Operation \"\(op.name)\" not allowed by SDK option allow.op value: \"\(allowOp)\"")
  }

  if op.points.isEmpty {
    throw ctx.makeError("point_no_points",
      "Operation \"\(op.name)\" has no endpoint definitions.")
  }

  if op.points.count == 1 {
    ctx.point = op.points[0]
  } else {
    let reqselector: VMap
    let selector: VMap
    if op.input == "data" {
      reqselector = ctx.reqdata
      selector = ctx.data
    } else {
      reqselector = ctx.reqmatch
      selector = ctx.match
    }

    var point: VMap? = nil
    for candidate in op.points {
      point = candidate
      let selectDef = gp(candidate, "select").asMap
      var found = true

      if let selectDef = selectDef {
        if let existList = gp(selectDef, "exist").asList {
          for ek in existList.items {
            let existkey = ek.asString ?? ""
            let rv = getprop(.map(reqselector), .string(existkey))
            let sv = getprop(.map(selector), .string(existkey))
            if isNil(rv) && isNil(sv) { found = false; break }
          }
        }
      }

      if found {
        let reqAction = gp(reqselector, "$action")
        let selectAction = gp(selectDef == nil ? .noval : .map(selectDef!), "$action")
        if reqAction != selectAction { found = false }
      }

      if found { break }
    }

    let reqAction = gp(reqselector, "$action")
    if !isNil(reqAction), let point = point {
      let pointSelect = gp(point, "select").asMap
      let pointAction = gp(pointSelect == nil ? .noval : .map(pointSelect!), "$action")
      if reqAction != pointAction {
        throw ctx.makeError("point_action_invalid",
          "Operation \"\(op.name)\" action \"\(stringify(reqAction))\" is not valid.")
      }
    }

    ctx.point = point
  }

  return ctx.point
}

func makeSpecUtil(_ ctx: Context) throws -> Spec {
  if let cached = ctx.out["spec"] as? Spec {
    ctx.spec = cached
    return cached
  }

  let options = ctx.options
  let utility = ctx.utility!

  let basev = gp(options, "base").asString ?? ""
  let prefix = gp(options, "prefix").asString ?? ""
  let suffix = gp(options, "suffix").asString ?? ""
  let parts = gp(ctx.point, "parts").asList

  let specmap = VMap()
  specmap.entries["base"] = .string(basev)
  specmap.entries["prefix"] = .string(prefix)
  specmap.entries["parts"] = parts == nil ? .noval : .list(parts!)
  specmap.entries["suffix"] = .string(suffix)
  specmap.entries["step"] = .string("start")

  ctx.spec = Spec(specmap)
  ctx.spec!.method = utility.prepareMethod(ctx)

  let allowMethod = gpath(options, "allow", "method").asString ?? ""
  if !allowMethod.contains(ctx.spec!.method) {
    throw ctx.makeError("spec_method_allow",
      "Method \"\(ctx.spec!.method)\" not allowed by SDK option allow.method value: \"\(allowMethod)\"")
  }

  ctx.spec!.params = utility.prepareParams(ctx)
  ctx.spec!.query = utility.prepareQuery(ctx)
  ctx.spec!.headers = utility.prepareHeaders(ctx)
  ctx.spec!.body = utility.prepareBody(ctx)
  ctx.spec!.path = utility.preparePath(ctx)

  if let explain = ctx.ctrl.explain {
    explain.entries["spec"] = .nat(ctx.spec!)
  }

  let spec = try utility.prepareAuth(ctx)
  ctx.spec = spec
  return spec
}

func makeUrlUtil(_ ctx: Context) throws -> String {
  guard let spec = ctx.spec else {
    throw ctx.makeError("url_no_spec", "Expected context spec property to be defined.")
  }
  guard let result = ctx.result else {
    throw ctx.makeError("url_no_result", "Expected context result property to be defined.")
  }

  var url = join(jtp(spec.base, spec.prefix, spec.path, spec.suffix), "/", true)
  let resmatch = VMap()

  for item in items(.map(spec.params)) {
    let key = item[0].asString ?? ""
    let val = item[1]
    if !isNil(val) {
      let pattern = "\\{" + escre(.string(key)) + "\\}"
      url = regexReplace(url, pattern, escurl(.string(stringify(val))))
      resmatch.entries[key] = val
    }
  }

  var qsep = "?"
  for item in items(.map(spec.query)) {
    let key = item[0].asString ?? ""
    let val = item[1]
    if !isNil(val) {
      url += qsep + escurl(.string(key)) + "=" + escurl(.string(stringify(val)))
      qsep = "&"
      resmatch.entries[key] = val
    }
  }

  result.resmatch = resmatch
  return url
}

func makeFetchDefUtil(_ ctx: Context) throws -> VMap {
  guard let spec = ctx.spec else {
    throw ctx.makeError("fetchdef_no_spec", "Expected context spec property to be defined.")
  }

  if ctx.result == nil { ctx.result = Result(nil) }

  spec.step = "prepare"

  let url = try ctx.utility!.makeUrl(ctx)
  spec.url = url

  let fetchdef = VMap()
  fetchdef.entries["url"] = .string(url)
  fetchdef.entries["method"] = .string(spec.method)
  fetchdef.entries["headers"] = .map(spec.headers)

  if !isNil(spec.body) {
    if spec.body.isMap {
      fetchdef.entries["body"] = .string(jsonify(spec.body, indent: 0))
    } else {
      fetchdef.entries["body"] = spec.body
    }
  }

  return fetchdef
}

func makeRequestUtil(_ ctx: Context) throws -> Response {
  if let cached = ctx.out["request"] as? Response { return cached }

  let utility = ctx.utility!

  var response = Response(nil)
  let result = Result(nil)
  ctx.result = result

  guard let spec = ctx.spec else {
    throw ctx.makeError("request_no_spec", "Expected context spec property to be defined.")
  }

  let fetchdef: VMap
  do {
    fetchdef = try utility.makeFetchDef(ctx)
  } catch {
    response.err = error
    ctx.response = response
    spec.step = "postrequest"
    return response
  }

  if let explain = ctx.ctrl.explain {
    explain.entries["fetchdef"] = .map(fetchdef)
  }

  spec.step = "prerequest"

  let url = gp(fetchdef, "url").asString ?? ""

  var fetched: Value = .noval
  var fetchErr: Error? = nil
  do {
    fetched = try utility.fetcher(ctx, url, fetchdef)
  } catch {
    fetchErr = error
  }

  if let fe = fetchErr {
    response.err = fe
  } else if isNil(fetched) {
    let m = VMap()
    m.entries["err"] = .nat(ctx.makeError("request_no_response", "response: undefined"))
    response = Response(m)
  } else if let fm = fetched.asMap {
    response = Response(fm)
  } else {
    response.err = ctx.makeError("request_invalid_response", "response: invalid type")
  }

  spec.step = "postrequest"
  ctx.response = response
  return response
}

func makeResponseUtil(_ ctx: Context) throws -> Response {
  if let cached = ctx.out["response"] as? Response { return cached }

  let utility = ctx.utility!

  guard let spec = ctx.spec else {
    throw ctx.makeError("response_no_spec", "Expected context spec property to be defined.")
  }
  guard let response = ctx.response else {
    throw ctx.makeError("response_no_response", "Expected context response property to be defined.")
  }
  guard let result = ctx.result else {
    throw ctx.makeError("response_no_result", "Expected context result property to be defined.")
  }

  spec.step = "response"

  _ = utility.resultBasic(ctx)
  _ = utility.resultHeaders(ctx)
  _ = utility.resultBody(ctx)
  _ = utility.transformResponse(ctx)

  if result.err == nil {
    result.ok = true
  }

  if let explain = ctx.ctrl.explain {
    explain.entries["result"] = .nat(result)
  }

  return response
}

func makeResultUtil(_ ctx: Context) throws -> Result {
  if let cached = ctx.out["result"] as? Result { return cached }

  let op = ctx.op!
  let entity = ctx.entity

  guard let spec = ctx.spec else {
    throw ctx.makeError("result_no_spec", "Expected context spec property to be defined.")
  }
  guard let result = ctx.result else {
    throw ctx.makeError("result_no_result", "Expected context result property to be defined.")
  }

  spec.step = "result"

  _ = ctx.utility!.transformResponse(ctx)

  if op.name == "list" {
    let resdata = result.resdata
    result.resdata = .list([])

    if let list = resdata.asList, list.items.count > 0, let entity = entity {
      let entities = VList()
      for entry in list.items {
        let ent = entity.make()
        if let entryMap = entry.asMap {
          ent.data(.map(entryMap))
        }
        entities.items.append(.nat(ent))
      }
      result.resdata = .list(entities)
    }
  }

  if let explain = ctx.ctrl.explain {
    explain.entries["result"] = .nat(result)
  }

  return result
}

File: ts/project/.sdk/tm/scala/utility/Make.scala
package SCALAPACKAGE.utility

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}
import java.util.Objects
import SCALAPACKAGE.core._
import SCALAPACKAGE.utility.struct.Struct

// makeError finalises a failed operation: wraps the causing error in an
// SdkError carrying the cleaned result and spec, records it on ctx.ctrl, and
// either throws it (default) or — when ctrl.throw is false — returns the
// result's fallback resdata instead.
object MakeError {
  def makeError(ctx0: Context, err0: RuntimeException): Object = {
    val ctx = if (ctx0 == null) new Context(new LinkedHashMap[String, Object](), null) else ctx0

    var opname = if (ctx.op == null) "" else ctx.op.name
    if ("" == opname || "_" == opname) opname = "unknown operation"

    var result = ctx.result
    if (result == null) result = new Result(new LinkedHashMap[String, Object]())
    result.ok = false

    var err = err0
    if (err == null) err = result.err
    if (err == null) err = ctx.makeError("unknown", "unknown error")

    val errmsg = if (err.getMessage == null) String.valueOf(err) else err.getMessage
    var msg = "ProjectNameSDK: " + opname + ": " + errmsg
    msg = Clean.clean(ctx, msg).asInstanceOf[String]

    result.err = null

    val spec = ctx.spec

    if (ctx.ctrl.explain != null) {
      val errRecord = new LinkedHashMap[String, Object]()
      errRecord.put("message", msg)
      ctx.ctrl.explain.put("err", errRecord)
    }

    var code = ""
    err match { case e: SdkError => code = e.code; case _ => }

    val sdkErr = new SdkError(code, msg, ctx)
    sdkErr.result = Clean.clean(ctx, result)
    sdkErr.spec = Clean.clean(ctx, spec)

    ctx.ctrl.err = sdkErr

    // Fire PreUnexpected so observability features (metrics, telemetry, audit,
    // debug) close/record error paths that never reach PreDone (e.g. a PrePoint
    // rbac short-circuit). Fires after ctx.ctrl.err is set so hooks can read the
    // error; features guard against double-recording when PreDone already fired.
    if (ctx.utility != null && ctx.utility.featureHook != null) {
      ctx.utility.featureHook(ctx, "PreUnexpected")
    }

    if (java.lang.Boolean.FALSE == ctx.ctrl.throwing) return result.resdata

    throw sdkErr
  }
}

object MakePoint {
  def makePoint(ctx: Context): JMap[String, Object] = {
    val outPoint = ctx.out.get("point")
    if (outPoint != null) {
      // A PrePoint feature hook (e.g. rbac) may short-circuit by storing an
      // error here; surface it before any endpoint resolution.
      outPoint match {
        case e: RuntimeException => throw e
        case m: JMap[_, _] => ctx.point = m.asInstanceOf[JMap[String, Object]]; return ctx.point
        case _ =>
      }
    }

    val op = ctx.op
    val options = ctx.options

    val allowOp = Struct.getpath(options, java.util.List.of("allow", "op")) match { case s: String => s; case _ => "" }
    if (!allowOp.contains(op.name)) {
      throw ctx.makeError("point_op_allow",
        "Operation \"" + op.name + "\" not allowed by SDK option allow.op value: \"" + allowOp + "\"")
    }

    if (op.points.isEmpty) {
      throw ctx.makeError("point_no_points", "Operation \"" + op.name + "\" has no endpoint definitions.")
    }

    if (op.points.size() == 1) {
      ctx.point = op.points.get(0)
    } else {
      val reqselector: JMap[String, Object] = if ("data" == op.input) ctx.reqdata else ctx.reqmatch
      val selector: JMap[String, Object] = if ("data" == op.input) ctx.data else ctx.matchData

      var point: JMap[String, Object] = null
      var i = 0
      var break = false
      while (i < op.points.size() && !break) {
        point = op.points.get(i)
        val selectDef = Helpers.toMapAny(Struct.getprop(point, "select"))
        var found = true

        if (selector != null && selectDef != null) {
          Struct.getprop(selectDef, "exist") match {
            case exist: JList[_] =>
              val eit = exist.asInstanceOf[JList[Object]].iterator()
              var ebreak = false
              while (eit.hasNext && !ebreak) {
                val existkey = eit.next() match { case s: String => s; case _ => "" }
                val rv = Struct.getprop(reqselector, existkey, null)
                val sv = Struct.getprop(selector, existkey, null)
                if (rv == null && sv == null) { found = false; ebreak = true }
              }
            case _ =>
          }
        }

        if (found) {
          val reqAction = Struct.getprop(reqselector, "$action", null)
          val selectAction = Struct.getprop(selectDef, "$action", null)
          if (!Objects.equals(reqAction, selectAction)) found = false
        }

        if (found) break = true else i += 1
      }

      if (reqselector != null) {
        val reqAction = Struct.getprop(reqselector, "$action", null)
        if (reqAction != null && point != null) {
          val pointSelect = Helpers.toMapAny(Struct.getprop(point, "select"))
          val pointAction = Struct.getprop(pointSelect, "$action", null)
          if (!Objects.equals(reqAction, pointAction)) {
            throw ctx.makeError("point_action_invalid",
              "Operation \"" + op.name + "\" action \"" + Struct.stringify(reqAction) + "\" is not valid.")
          }
        }
      }

      ctx.point = point
    }

    ctx.point
  }
}

object MakeOptions {
  def makeOptions(ctx: Context): JMap[String, Object] = {
    var options = ctx.options
    if (options == null) options = new LinkedHashMap[String, Object]()

    // Merge custom utility overrides onto the utility object.
    val customUtils = Helpers.toMapAny(options.get("utility"))
    if (customUtils != null && ctx.utility != null) ctx.utility.custom.putAll(customUtils)

    var opts = Struct.clone(options).asInstanceOf[JMap[String, Object]]

    // Feature add-order. options.feature may be given as an ordered LIST of
    // { name, active, ...opts } entries (the list position IS the order in
    // which features are added), or as a { name: {opts} } map. Normalize a
    // list to a map (so merge/validate/init are unchanged) and remember the
    // explicit order; a map defaults to test-first so the `test` mock
    // transport is installed as the base of the transport wrapper chain.
    val featureorder = new ArrayList[Object]()
    opts.get("feature") match {
      case flist: JList[_] =>
        val fmap = new LinkedHashMap[String, Object]()
        val fit = flist.asInstanceOf[JList[Object]].iterator()
        while (fit.hasNext) {
          val em = Helpers.toMapAny(fit.next())
          if (em != null) {
            em.get("name") match {
              case fname: String if "" != fname =>
                val fopts = new LinkedHashMap[String, Object](em)
                fopts.remove("name")
                fmap.put(fname, fopts)
                featureorder.add(fname)
              case _ =>
            }
          }
        }
        opts.put("feature", fmap)
      case _ =>
    }

    var config = ctx.config
    if (config == null) config = new LinkedHashMap[String, Object]()
    var cfgopts = Helpers.toMapAny(config.get("options"))
    if (cfgopts == null) cfgopts = new LinkedHashMap[String, Object]()

    val optspec = Json.parse(
      "{"
        + "\"apikey\": \"\","
        + "\"base\": \"http://localhost:8000\","
        + "\"prefix\": \"\","
        + "\"suffix\": \"\","
        + "\"auth\": { \"prefix\": \"\" },"
        + "\"headers\": { \"`$CHILD`\": \"`$STRING`\" },"
        + "\"allow\": {"
        + "  \"method\": \"GET,PUT,POST,PATCH,DELETE,OPTIONS\","
        + "  \"op\": \"create,update,load,list,remove,command,direct\""
        + "},"
        + "\"entity\": { \"`$CHILD`\": {"
        + "  \"`$OPEN`\": true, \"active\": false, \"alias\": {} } },"
        + "\"feature\": { \"`$CHILD`\": {"
        + "  \"`$OPEN`\": true, \"active\": false } },"
        + "\"utility\": {},"
        + "\"system\": {},"
        + "\"test\": { \"active\": false, \"entity\": { \"`$OPEN`\": true } },"
        + "\"clean\": { \"keys\": \"key,token,id\" }"
        + "}").asInstanceOf[JMap[String, Object]]

    // Preserve system.fetch before merge/validate.
    var sysFetch = Struct.getpath(opts, java.util.List.of("system", "fetch"))
    if (sysFetch eq Struct.UNDEF) sysFetch = null

    val mergeList = new ArrayList[Object]()
    mergeList.add(new LinkedHashMap[String, Object]())
    mergeList.add(cfgopts)
    mergeList.add(opts)
    val merged = Struct.merge(mergeList)

    val vopts = new LinkedHashMap[String, Object]()
    vopts.put("errs", new ArrayList[Object]())
    val validated = Struct.validate(merged, optspec, vopts)
    opts = validated.asInstanceOf[JMap[String, Object]]

    // Restore system.fetch.
    if (sysFetch != null) {
      val sys = Helpers.toMapAny(opts.get("system"))
      if (sys != null) sys.put("fetch", sysFetch)
      else { val sm = new LinkedHashMap[String, Object](); sm.put("fetch", sysFetch); opts.put("system", sm) }
    }

    // Derived clean config.
    var cleanKeys = "key,token,id"
    Struct.getpath(opts, java.util.List.of("clean", "keys")) match { case s: String => cleanKeys = s; case _ => }

    val filtered = new ArrayList[String]()
    for (p0 <- cleanKeys.split(",")) {
      val p = p0.trim
      if ("" != p) filtered.add(Struct.escre(p))
    }
    val keyre = String.join("|", filtered)

    // Resolve the feature add-order: an explicit list order (above) wins;
    // otherwise order the map test-first, then the remaining names sorted, so
    // the outcome is deterministic and `test` is always the base transport.
    if (featureorder.isEmpty) {
      val fmap = Helpers.toMapAny(opts.get("feature"))
      val names = new ArrayList[String]()
      if (fmap != null) names.addAll(fmap.keySet())
      java.util.Collections.sort(names)
      if (names.contains("test")) {
        featureorder.add("test")
        val nit = names.iterator()
        while (nit.hasNext) { val n = nit.next(); if ("test" != n) featureorder.add(n) }
      } else {
        featureorder.addAll(names)
      }
    }

    val derived = new LinkedHashMap[String, Object]()
    val derivedClean = new LinkedHashMap[String, Object]()
    if ("" != keyre) derivedClean.put("keyre", keyre)
    derived.put("clean", derivedClean)
    derived.put("featureorder", featureorder)
    opts.put("__derived__", derived)

    opts
  }
}

object MakeFetchDef {
  def makeFetchDef(ctx: Context): JMap[String, Object] = {
    val spec = ctx.spec
    if (spec == null) throw ctx.makeError("fetchdef_no_spec", "Expected context spec property to be defined.")

    if (ctx.result == null) ctx.result = new Result(new LinkedHashMap[String, Object]())

    spec.step = "prepare"

    val url = ctx.utility.makeUrl(ctx)
    spec.url = url

    val fetchdef = new LinkedHashMap[String, Object]()
    fetchdef.put("url", url)
    fetchdef.put("method", spec.method)
    fetchdef.put("headers", spec.headers)

    if (spec.body != null) {
      spec.body match {
        case _: JMap[_, _] => fetchdef.put("body", Struct.jsonify(spec.body))
        case _ => fetchdef.put("body", spec.body)
      }
    }

    fetchdef
  }
}

object MakeRequest {
  def makeRequest(ctx: Context): Response = {
    ctx.out.get("request") match { case r: Response => return r; case _ => }

    val spec = ctx.spec
    val utility = ctx.utility

    var response = new Response(new LinkedHashMap[String, Object]())
    val result = new Result(new LinkedHashMap[String, Object]())
    ctx.result = result

    if (spec == null) throw ctx.makeError("request_no_spec", "Expected context spec property to be defined.")

    var fetchdef: JMap[String, Object] = null
    try fetchdef = utility.makeFetchDef(ctx)
    catch {
      case err: RuntimeException =>
        response.err = err; ctx.response = response; spec.step = "postrequest"; return response
    }

    if (ctx.ctrl.explain != null) ctx.ctrl.explain.put("fetchdef", fetchdef)

    spec.step = "prerequest"

    val url = fetchdef.get("url")
    var fetched: Object = null
    var fetchErr: RuntimeException = null
    try fetched = utility.fetcher(ctx, url match { case s: String => s; case _ => "" }, fetchdef)
    catch { case err: RuntimeException => fetchErr = err }

    if (fetchErr != null) response.err = fetchErr
    else if (fetched == null) {
      val resmap = new LinkedHashMap[String, Object]()
      resmap.put("err", ctx.makeError("request_no_response", "response: undefined"))
      response = new Response(resmap)
    } else fetched match {
      case m: JMap[_, _] => response = new Response(m.asInstanceOf[JMap[String, Object]])
      case _ => response.err = ctx.makeError("request_invalid_response", "response: invalid type")
    }

    spec.step = "postrequest"
    ctx.response = response
    response
  }
}

object MakeResponse {
  def makeResponse(ctx: Context): Response = {
    ctx.out.get("response") match { case r: Response => return r; case _ => }

    val utility = ctx.utility
    val spec = ctx.spec
    val result = ctx.result
    val response = ctx.response

    if (spec == null) throw ctx.makeError("response_no_spec", "Expected context spec property to be defined.")
    if (response == null) throw ctx.makeError("response_no_response", "Expected context response property to be defined.")
    if (result == null) throw ctx.makeError("response_no_result", "Expected context result property to be defined.")

    spec.step = "response"

    utility.resultBasic(ctx)
    utility.resultHeaders(ctx)
    utility.resultBody(ctx)
    utility.transformResponse(ctx)

    if (result.err == null) result.ok = true

    if (ctx.ctrl.explain != null) ctx.ctrl.explain.put("result", result)

    response
  }
}

object MakeResult {
  def makeResult(ctx: Context): Result = {
    ctx.out.get("result") match { case r: Result => return r; case _ => }

    val utility = ctx.utility
    val op = ctx.op
    val entity = ctx.entity
    val spec = ctx.spec
    val result = ctx.result

    if (spec == null) throw ctx.makeError("result_no_spec", "Expected context spec property to be defined.")
    if (result == null) throw ctx.makeError("result_no_result", "Expected context result property to be defined.")

    spec.step = "result"

    utility.transformResponse(ctx)

    if ("list" == op.name) {
      val resdata = result.resdata
      result.resdata = new ArrayList[Object]()

      resdata match {
        case l: JList[_] if !l.isEmpty && entity != null =>
          val entities = new ArrayList[Object]()
          val it = l.asInstanceOf[JList[Object]].iterator()
          while (it.hasNext) {
            val entry = it.next()
            val ent = entity.make()
            entry match { case m: JMap[_, _] => ent.data(m); case _ => }
            entities.add(ent)
          }
          result.resdata = entities
        case _ =>
      }
    }

    if (ctx.ctrl.explain != null) ctx.ctrl.explain.put("result", result)

    result
  }
}

object MakeUrl {
  def makeUrl(ctx: Context): String = {
    val spec = ctx.spec
    val result = ctx.result

    if (spec == null) throw ctx.makeError("url_no_spec", "Expected context spec property to be defined.")
    if (result == null) throw ctx.makeError("url_no_result", "Expected context result property to be defined.")

    val joinParts = new ArrayList[Object]()
    joinParts.add(spec.base)
    joinParts.add(spec.prefix)
    joinParts.add(spec.path)
    joinParts.add(spec.suffix)
    var url = Struct.join(joinParts, "/", true)

    val resmatch = new LinkedHashMap[String, Object]()

    val pit = Struct.items(spec.params).iterator()
    while (pit.hasNext) {
      val item = pit.next()
      val key = item.get(0) match { case s: String => s; case _ => "" }
      val v = item.get(1)
      if (v != null) {
        url = url.replaceAll("\\{" + Struct.escre(key) + "\\}",
          java.util.regex.Matcher.quoteReplacement(Struct.escurl(Struct.stringify(v))))
        resmatch.put(key, v)
      }
    }

    var qsep = "?"
    val qit = Struct.items(spec.query).iterator()
    while (qit.hasNext) {
      val item = qit.next()
      val key = item.get(0) match { case s: String => s; case _ => "" }
      val v = item.get(1)
      if (v != null) {
        url += qsep + Struct.escurl(key) + "=" + Struct.escurl(Struct.stringify(v))
        qsep = "&"
        resmatch.put(key, v)
      }
    }

    result.resmatch = resmatch
    url
  }
}
